cnd/src/asset/ethereum/erc20.rs

```diff
@@ -0,0 +1,190 @@
+    }
+
+    pub fn to_u256(&self) -> U256 {
+        let buf = self.0.to_bytes_be();
+        U256::from_big_endian(&buf)
+    }
+}
+
+impl FromWei<U256> for Erc20Quantity {
+    fn from_wei(wei: U256) -> Self {
+        let mut buf = [0u8; 32];
+        wei.to_big_endian(&mut buf);
+        Self(BigUint::from_bytes_be(&buf))
+    }
+}
+
+impl fmt::Display for Erc20Quantity {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+macro_rules! impl_from_wei_primitive {
+    ($primitive:ty) => {
+        impl FromWei<$primitive> for Erc20Quantity {
+            fn from_wei(w: $primitive) -> Self {
+                Erc20Quantity(BigUint::from(w))
+            }
+        }
+    };
+}
+
+impl_from_wei_primitive!(u8);
+impl_from_wei_primitive!(u16);
+impl_from_wei_primitive!(u32);
+impl_from_wei_primitive!(u64);
+impl_from_wei_primitive!(u128);
+
+impl TryFromWei<BigUint> for Erc20Quantity {
+    type Err = Error;
+    fn try_from_wei(wei: BigUint) -> Result<Self, Self::Err> {
+        if wei > Self::max_value().0 {
+            Err(Error::Overflow)
+        } else {
+            Ok(Self(wei))
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Erc20Quantity {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as Deserializer<'de>>::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct Visitor;
+
+        impl<'vde> de::Visitor<'vde> for Visitor {
+            type Value = Erc20Quantity;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+                formatter.write_str("A string representing an ERC20 quantity")
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Erc20Quantity, E>
+            where
+                E: de::Error,
+            {
+                let wei = BigUint::from_str(v).map_err(E::custom)?;
+                let quantity = Erc20Quantity::try_from_wei(wei).map_err(E::custom)?;
+                Ok(quantity)
+            }
+        }
+
+        deserializer.deserialize_str(Visitor)
+    }
+}
+
+impl Serialize for Erc20Quantity {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.to_string().as_str())
+    }
+}
+
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct Erc20 {
+    pub token_contract: Address,
+    pub quantity: Erc20Quantity,
+}
+
+impl fmt::Display for Erc20 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.quantity)
+    }
+}
+
+impl Erc20 {
+    pub fn new(token_contract: Address, quantity: Erc20Quantity) -> Self {
+        Erc20 {
+            token_contract,
+            quantity,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn from_one_thousand_in_u256_equals_one_thousand_u32() {
+        let u256 = U256::from(1000);
```

Comment:
:wrench: And in other places.

```suggestion
        let u256 = U256::from(1_000);
```